File: 01-intro/03-basic-combinatorial-algorithms/04-sum-combinations/sum-combinations.py
from pprint import pprint as print
from sys import argv


class SumCombinations(object):
    def __init__(self, number: int):
        self.number = number
        self.sums = []
        self.buffer = {}
    # def __init__

    def _compute(self, start: int, position: int):
        if start == 0:
            self.sums.append(list(self.buffer.values())[1:])
            return
        # if

        for head in range(start, 0, -1):
            self.buffer[position] = head
            if self.buffer[position] <= self.buffer[position - 1]:
                self._compute(start - head, position + 1)
            # if
        # for

        return self
    # def _compute

    def compute(self):
        self.buffer[0] = self.number + 1
        self._compute(self.number, 1)

        return self
    # def compute
# class SumCombinations

UPPER_LIMIT = 4 if len(argv) == 1 else (int(argv[1]) + 1)
INIT_INDEX = 1
for i in range(INIT_INDEX, UPPER_LIMIT):
    print({i: SumCombinations(i).compute().sums})

File: 01-intro/03-basic-combinatorial-algorithms/04-sum-combinations/sum-combinations.ex
defmodule SumCombinations do
  def compute(n), do: compute(n, 1)

  def compute(n, pos), do: compute(n, pos, %{0 => n + 1})

  def compute(n, pos, buffer) do
    if n == 0 do
      print_buffer(buffer)
    else
      for k <- n..1 do
        buffer = Map.merge(buffer, %{pos => k})

        if(buffer[pos] <= buffer[pos - 1]) do
          compute(n - k, pos + 1, buffer)
        end
      end
    end
  end

  def print_buffer(buffer) do
    [head | tail] = Map.values(buffer)
    IO.inspect(%{(head - 1) => tail}, label: "combination")
  end
end

for i <- 0..6, do: SumCombinations.compute(i)
